src/libraries/Common/tests/System/Net/Sockets/SocketTestExtensions.cs

```diff
@@ -49,5 +51,30 @@ public static (Socket, Socket) CreateConnectedSocketPair()
 
             return (client, server);
         }
+
+        // Tries to connect within the provided timeout interval
+        // Useful to speed up "can not connect" assertions on Windows
+        public static bool TryConnect(this Socket socket, EndPoint remoteEndpoint, int millisecondsTimeout)
+        {
+            // To avoid race conditions, ManualResetEventSlim is left undisposed,
+            // letting SafeHandle's finalizer do the cleanup work, if needed
+            var mre = new ManualResetEventSlim(false);
+            using var sea = new SocketAsyncEventArgs()
+            {
+                RemoteEndPoint = remoteEndpoint,
+                UserToken = mre
+            };
+
+            sea.Completed += (s, e) => ((ManualResetEventSlim)e.UserToken).Set();
+
+            bool pending = socket.ConnectAsync(sea);
+            if (!pending || mre.Wait(millisecondsTimeout))
+            {
```

Comment:
It's not critical, but you could dispose of `mre` inside this if block if you wanted to.